source/common/uma_pools/usm_allocator.cpp

```diff
@@ -0,0 +1,866 @@
+// USM allocation. Thus, for example, for a 64KB minimum allocation size,
+// and 8-byte allocations, only 1 in ~8000 requests results in a new
+// USM allocation. Freeing results only in a chunk of a larger allocation
+// to be marked as available and no real return to the system.
+// An allocation is returned to the system only when all
+// chunks in the larger allocation are freed by the program.
+// Allocations larger than ChunkCutOff use a separate USM allocation for each
+// request. These are subject to "pooling". That is, when such an allocation is
+// freed by the program it is retained in a pool. The pool is available for
+// future allocations, which means there are fewer actual USM
+// allocations/deallocations.
+
+// The largest size which is allocated via the allocator.
+// Allocations with size > CutOff bypass the USM allocator and
+// go directly to the runtime.
+static constexpr size_t CutOff = (size_t)1 << 31; // 2GB
+
+// Aligns the pointer down to the specified alignment
+// (e.g. returns 8 for Size = 13, Alignment = 8)
+static void *AlignPtrDown(void *Ptr, const size_t Alignment) {
+  return reinterpret_cast<void *>((reinterpret_cast<size_t>(Ptr)) &
+                                  (~(Alignment - 1)));
+}
+
+// Aligns the pointer up to the specified alignment
+// (e.g. returns 16 for Size = 13, Alignment = 8)
+static void *AlignPtrUp(void *Ptr, const size_t Alignment) {
+  void *AlignedPtr = AlignPtrDown(Ptr, Alignment);
+  // Special case when the pointer is already aligned
+  if (Ptr == AlignedPtr) {
+    return Ptr;
+  }
+  return static_cast<char *>(AlignedPtr) + Alignment;
+}
+
+// Aligns the value up to the specified alignment
+// (e.g. returns 16 for Size = 13, Alignment = 8)
+static size_t AlignUp(size_t Val, size_t Alignment) {
+  assert(Alignment > 0);
+  return (Val + Alignment - 1) & (~(Alignment - 1));
+}
+
+class Bucket;
+
+// Represents the allocated memory block of size 'SlabMinSize'
+// Internally, it splits the memory block into chunks. The number of
+// chunks depends of the size of a Bucket which created the Slab.
+// The chunks
+// Note: Bucket's method are responsible for thread safety of Slab access,
+// so no locking happens here.
+class Slab {
+
+  // Pointer to the allocated memory of SlabMinSize bytes
+  void *MemPtr;
+
+  // Represents the current state of each chunk:
+  // if the bit is set then the chunk is allocated
+  // the chunk is free for allocation otherwise
+  std::vector<bool> Chunks;
+
+  // Total number of allocated chunks at the moment.
+  size_t NumAllocated = 0;
+
+  // The bucket which the slab belongs to
+  Bucket &bucket;
+
+  using ListIter = std::list<std::unique_ptr<Slab>>::iterator;
+
+  // Store iterator to the corresponding node in avail/unavail list
+  // to achieve O(1) removal
+  ListIter SlabListIter;
+
+  // Hints where to start search for free chunk in a slab
+  size_t FirstFreeChunkIdx = 0;
+
+  // Return the index of the first available chunk, -1 otherwize
```

Comment:
`-1` will never be returned as the type returned is `size_t`. Perhaps ssize_t should be used instead. Otherwise this return value should be documented as SIZE_MAX and it would be better to return SIZE_MAX explicitly in this method's implementation rather than implicitly returning it from `-1` value.

Reply:
Done.